File: Binary_Tree_Right_Side_View_199.py

# Time- O(n)
# Space -O(n)
from collections import deque
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if root is None:
            return None
        ans = []
        
        q =deque()
        
        q.append(root)

        while q:
            
            level_size = len(q)
            
            for _ in range(level_size):
                curr = q.popleft()
                temp = curr.val                
                if curr.left:
                    q.append(curr.left)                    
                if curr.right:
                    q.append(curr.right)
            ans.append(temp)
            
        return ans

File: Binary_Tree_Right_Side_View_199.java
//Time Complexity : O(n)
//Space Complexity : O(n)
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : None

package com.s30.satish;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;


class Binary_Tree_Right_Side_View_199 {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if(root == null)
            return result;
        
        Queue<TreeNode> bfsQueue = new LinkedList<>();
        bfsQueue.add(root);
        
        while(!bfsQueue.isEmpty())
        {
            int size = bfsQueue.size();
            int last = -1;
            for(int i = 0; i < size; i++)
            {
                TreeNode node = bfsQueue.poll();

                if(i == size-1)
                	last = node.val;
                if(node.left != null)
                    bfsQueue.add(node.left);
                if(node.right != null)
                    bfsQueue.add(node.right);
            }
            result.add(last);
        }
        return result;        
    }
}
